File: find_right_interval.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<pair<int, int>> starts;
    int N;
    
    int fn(int x) {
        int lo = 0;
        int hi = N - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int y = starts[mid].first;
            
            if ((mid == 0 || starts[mid - 1].first < x) && starts[mid].first >= x) {
                return starts[mid].second;
            }
            
            if (starts[mid].first >= x) {
                hi = mid - 1;   
            } else {
                lo = mid + 1;
            }
        }
        return -1;            
    }
    
    vector<int> findRightInterval(vector<Interval>& intervals) {
        N = intervals.size();
        for (int i = 0; i < N; i++) {
            starts.push_back({intervals[i].start, i});
        }
        sort(starts.begin(), starts.end());
        
        vector<int> res;
        for (auto interval: intervals) {
            int i = fn(interval.end);
            res.push_back(i);
        }
        return res;
    }
};

File: find-right-interval.java
class StartAndIndex implements Comparable<StartAndIndex> {
    int start;
    int index;
    public StartAndIndex(int s, int i) {
        start = s;
        index = i;
    }
    
    public int compareTo(StartAndIndex other) {
        if (other.start < start || (other.start == start && other.index < index)) {
            return -1;
        } else if (other.start == start && other.index == index) {
            return 0;
        } else {
            return 1;
        }
    }
}

class Solution {
    private int findIndexOfGreaterThan(StartAndIndex[] A, int t) {
        int best = -1;
        int lo = 0;
        int hi = A.length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (A[mid].start >= t) {
                best = A[mid].index;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        
        return best;
    }
    
    public int[] findRightInterval(int[][] intervals) {
        int n = intervals.length;
        StartAndIndex[] starts = new StartAndIndex[n];
        for (int i = 0; i < n; i++) {
            starts[i] = new StartAndIndex(intervals[i][0], i);
        }
        Arrays.sort(starts);
        
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            res[i] = findIndexOfGreaterThan(starts, intervals[i][1]);
        }
        
        return res;
    }
}
